### PR TITLE
feat(react): Update stable types for v17

### LINKED ISSUE #43985: [@types/react] add enterkeyhint attribute to input and textarea
If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/react` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @johnnyreilly @bbenezech @pzavolinsky @digiguru @ericanderson @DovydasNavickas @theruther4d @guilhermehubner @ferdaber @jrakotoharisoa @pascaloliv @hotell @franklixuefei @Jessidhia @saranshkataria @lukyth @eps1lon @zieka @dancerphil @dimitropoulos @disjukr @vhfmag

There is experimental input and textarea attribute `enterkeyhint` https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes that does not seem to be supported by type definitions. Even though experimental it is supported on latest releases of safari and chrome, perhaps it would be a good idea to add it?

Currently following error pops up when trying to use it in react:

Screenshot 2020-04-17 at 14 57 06

### PR DESCRIPTION
**Pull request by eps1lon**
Closes #43985

~Not including updates for `experimental` release channel. Goal is to have a minimal update that can be shipped as soon as possible.~ Experimental release channel handled in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/49152

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
  - https://github.com/facebook/react/releases/tag/v17.0.0
  - https://html.spec.whatwg.org/multipage/interaction.html#attr-enterkeyhint
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- ~[ ]~ If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react-dom ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-dom@latest/))
was missing the following properties:
1. flushSync
1. unstable_createPortal
## react-is ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-is@latest/))
was missing the following properties:
1. isConcurrentMode
1. isConcurrentMode
1. isConcurrentMode
## react-test-renderer ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-test-renderer@latest/))
was missing the following properties:
1. unstable_batchedUpdates
1. unstable_concurrentAct
1. unstable_batchedUpdates
1. unstable_concurrentAct
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 3d3fb93afde7cd08afadb558fea4412e5beb93f5
</p>

**Comment by eps1lon** [comment]
Looks like there's just to much to benchmark.

File: types/react/experimental.d.ts
Comment (maraisr) [inline types/react/experimental.d.ts:165]: Do we need to change types here now that `timeoutMs` is removed?

see https://github.com/facebook/react/pull/19703

Comment (eps1lon) [reply types/react/experimental.d.ts:165]: Not including updates for experimental release channel. Goal is to have a minimal update that can be shipped as soon as possible.

**Comment by dispix** [comment]
I think this PR should take the time to fix a long running issue with the `FC` type: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/40993

Having an optional `children` by default doesn't make sense but since it's a breaking change, removing it in this PR would make sense.

**Comment by eps1lon** [comment]
> I think this PR should take the time to fix a long running issue with the `FC` type: #40993
> 
> Having an optional `children` by default doesn't make sense but since it's a breaking change, removing it in this PR would make sense.

We've discussed this in https://github.com/DefinitelyTyped/DefinitelyTyped/issues/46691 and decided to postpone any type related breaking change to v18 because React core members specifically asked us not to. If you disagree please continue the discussion in the issue.

**Comment by dispix** [comment]
> > I think this PR should take the time to fix a long running issue with the `FC` type: #40993
> > Having an optional `children` by default doesn't make sense but since it's a breaking change, removing it in this PR would make sense.
> 
> We've discussed this in #46691 and decided to postpone any type related breaking change to v18 because React core members specifically asked us not to. If you disagree please continue the discussion in the issue.

Oh, I missed that one but that makes sense. Thanks for the info :+1:

**Comment by alecmev** [comment]
[`disableRemotePlayback`](https://github.com/facebook/react/pull/18619/files) might be a safe addition?

File: types/react/experimental.d.ts
Comment (eps1lon) [reply types/react/experimental.d.ts:165]: @maraisr I suspect this PR takes longer since a maintainer has to manually unblock it (benchmark failing due to timeout) so I opened https://github.com/DefinitelyTyped/DefinitelyTyped/pull/49152 for `@experimental`.

**Comment by eps1lon** [comment]
> [`disableRemotePlayback`](https://github.com/facebook/react/pull/18619/files) might be a safe addition?

@alecmev `disableRemotePlayback` is included in this PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/48971/files#diff-32cfd8cb197872bcba371f5018185d2e75fa540b52cda2dd7d8ac12dcc021299R2360

File: types/react/experimental.d.ts
Comment (eps1lon) [reply types/react/experimental.d.ts:165]: Released in `@types/react@16.9.54`

**Review by n1ru4l — CHANGES_REQUESTED** [review]
With TypeScript 4.1.x and create-react-app (or any other situation that enforces/uses jsx-runtime)  you will get the following error when trying to use react:

```
Could not find a declaration file for module 'react/jsx-runtime'. '.../node_modules/react/jsx-runtime.js' implicitly has an 'any' type.
  If the 'react' package actually exposes this module, consider sending a pull request to amend 'https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react`  TS7016
```

It seems like a type definition for `react/jsx-runtime` is missing ([related file in original react repository](https://github.com/facebook/react/blob/89d4fe141a28b54807901126063249cab6483f89/packages/react/jsx-runtime.js#L1-L9))

Related issue: https://github.com/facebook/create-react-app/issues/10109

**Comment by eps1lon** [comment]
@n1ru4l Why did you request changes? The PR is already open for a month and now postponed because it doesn't include one feature you want. We could've fixed this in a separate PR but now everybody has to wait longer.

**Comment by n1ru4l** [comment]
@eps1lon No hard feelings, I just think it is crucial to have this in here because otherwise, it react is unusable for most people that are using react or create-react-app with TypeScript 4.1 and do not know how to add the definitions manually.

I did not request the change for myself but for the community.

Anyways, it should be straight-forward, as the file only exports 3 things... Let me know if you need help with it or "dismiss" my feedback as the bot suggests..

**Comment by eps1lon** [comment]
It could've been a separate PR. This PR is valid without the runtime. You could've made that change based on this PR but instead you blocked everyone even if they don't use the jsx-runtime.

**Comment by n1ru4l** [comment]
Just dismiss it instead of blaming me then 😅

File: types/react/jsx-dev-runtime.d.ts
Comment (Andarist) [inline types/react/jsx-dev-runtime.d.ts:1]: Those files should export `JSX` namespace - you can look at how I have done it for Emotion here: https://github.com/emotion-js/emotion/pull/2078/files . 

It would probably help if the `JSX` namespace in the `index` would actually be exported somehow rather than just attached to the global namespace. I've attempted to do a similar refactor here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46351 but I didn't receive the required approvals. You can maybe incorporate part of this work into your branch.

File: types/react/jsx-runtime.d.ts
Comment (eps1lon) [inline types/react/jsx-runtime.d.ts:3]: That's almost certainly not correct but at least it gets some type-checking. We'll have to figure out later what the correct approach would be. Personally, I would've preferred we introduce it in a separate PR (since older majors also need it) but otherwise the PR is blocked.

**Comment by eps1lon** [comment]
> Just dismiss it instead of blaming me then  - Also, I suggested changes not forced you to apply them...

We have to be a bit more considerate with the feedback we give in an automated repository.

File: types/react/jsx-dev-runtime.d.ts
Comment (eps1lon) [reply types/react/jsx-dev-runtime.d.ts:1]: Can we fix this in a follow-up? The current approach at least fixed https://github.com/facebook/create-react-app/issues/10109 and has less risk of breaking existing patterns. As far as I know the jsx `jsx-runtime` and `createElement` allow are the same structurally.

File: types/react/jsx-runtime.d.ts
Comment (Andarist) [reply types/react/jsx-runtime.d.ts:3]: If you plan to work on those entries as part of other PR I would rather recommend removing this entirely - at least the `createElement` export because as you have said: it's not correct and just won't match the expectations.

I'm not 100% sure about the TS resolution algorithm because they are somewhat magic to me but what it might do right now is just pulling the root entry (`index`) and thus bringing those types into the project (the global ones) that might make its global JSX namespace work. OTOH those types should probably be already pulled by other means so not sure if this is at all needed here (if you plan to work on this in a separate PR).

**Comment by Andarist** [comment]
> We have to be a bit more considerate with the feedback we give in an automated repository.

I must say that I agree with @n1ru4l on this one. His intentions were good and coming from the care about the users and various use cases, and bashing him like this is not helpful. He could also not be fully aware of all the implications of his actions on the automated process. It should be enough to just comment back on it explaining the problem without dragging this through several comments that are helpful to anyone.

File: types/react/jsx-dev-runtime.d.ts
Comment (eps1lon) [reply types/react/jsx-dev-runtime.d.ts:1]: > Those files should export JSX namespace - you can look at how I have done it for Emotion here: emotion-js/emotion#2078 (files) .

As far as I can tell `import './index'` works as well since that `./index` adds the JSX namespace globally.

This seems safer to me and any refactoring we want to do to how we handle JSX namespaces can be done in a separate PR. Or would this block emotion? If so could you propose a change to this branch so that I can understand what exactly you're trying to do?

**Comment by n1ru4l** [comment]
Having only contributed sparingly to definitely typed I am honestly not totally aware of this process. Anyway, I don't see an issue as the bot clearly mentions you can just dismiss it? 

Screenshot 2020-11-20 at 12 03 53

Honestly, I think this behaviour of blaming is pretty toxic and will scare people away from making meaningful contributions...

**Comment by eps1lon** [comment]
> His intentions were good

I never assume something else. What matters is the impact.

> He could also not be fully aware of all the implications of his actions on the automated process. 

Now he is. We'll see if this changes his actions.

> bashing him like this is not helpful

If that was bashing then how is bashing me helpful? 

> Anyway, I don't see an issue as the bot clearly mentions you can just dismiss it?

Collaborators can which I'm not.

> Honestly, I think this behaviour of blaming is pretty toxic and will scare people away from making meaningful contributions...

Considering you had no problem coming in and requesting changes I don't see how you're easily scared. Just open an issue, explain your problem and we go from there. But demanding a certain implementation is  certainly the more toxic behavior.

File: types/react/jsx-dev-runtime.d.ts
Comment (Andarist) [reply types/react/jsx-dev-runtime.d.ts]: > Can we fix this in a follow-up? The current approach at least fixed facebook/create-react-app#10109 and has less risk of breaking existing patterns.

My intuition (but I haven't fully confirmed it) is that it's exactly what you have mentioned in the later post - `import './index'` does the trick and not the `createElement` part. So I'm just proposing to sticking to the first one because exporting `createElement` is not correct here

> As far as I know the jsx jsx-runtime and createElement allow are the same structurally.

Not really - `createElement` accepts variadic length of parameters while `jsx` and `jsxs` have fixed length of those. Those signatures do not match.

> Or would this block emotion?

Can't be totally sure, because how TS resolves those things is magic to me. My bet is that it doesn't affect us much because the global `JSX` namespace should be overshadowed by the one that we provide - so I think this is OK.

OTOH I believe that there is no need at all to provide a global JSX namespace and it's probably an artifact of the past and this (new major version of those types here) could be a good moment to clean this up.

**Comment by Andarist** [comment]
> If that was bashing then how is bashing me helpful?

I was trying to stand up for a fellow OSS contributor - because I think he had more right here and I didn't want to stay silent about this. Just trying to bring my 2 cents here so you could maybe see an alternative perspective and reconsider if how you have handled this was appropriate or not.

If you have considered the words used by me as bashing to you, please take my sincere apologies. I'm not a native speaker so choosing the appropriate, polite, wording is not my strongest suit. My intention was not to bash you but to stand up for @n1ru4l .

File: types/react/jsx-dev-runtime.d.ts
Comment (eps1lon) [reply types/react/jsx-dev-runtime.d.ts]: > Not really - createElement accepts variadic length of parameters while jsx and jsxs have fixed length of those. Those signatures do not match.

I'm specifically talking about the JSX. Not how it's transpiled. Since you don't "manually call"  `jsx` I don't see an issue.

> OTOH I believe that there is no need at all to provide a global JSX namespace and it's probably an artifact of the past and this (new major version of those types here) could be a good moment to clean this up.

I'm not doing feature changes and refactorings of existing systems in the same PR. Especially if I'm not aware of a concrete issue the current approach has. 

So it would help a lot if we talk about concrete issues (ideally that we can reproduce) instead of theoretical concerns. There are plenty of issues with the react typings that we didn't fix with 17 because we were [explicitly asked to keep 17 free from breaking changes](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/46691#issuecomment-672840456). And that's my goal: provide typings for React 17 that don't interfere with React.

Comment (Andarist) [reply types/react/jsx-dev-runtime.d.ts]: > I'm specifically talking about the JSX. Not how it's transpiled. Since you don't "manually call" jsx I don't see an issue.

That's fine by me and yes - JSX namespace works the same for both runtimes. My sole point is that this `createElement` should not appear in those new entries because it doesn't match reality and I don't think it's solving anything (unless it is but I somewhat doubt it - I've been tinkering lately with fixing JSX namespace lookup in TS and I don't see how this would be required here, but again - the whole thing is quirky to me so I might be off about some details).

> So it would help a lot if we talk about concrete issues (ideally that we can reproduce) instead of theoretical concerns. There are plenty of issues with the react typings that we didn't fix with 17 because we were explicitly asked to keep 17 free from breaking changes. And that's my goal: provide typings for React 17 that don't interfere with React.

👍

**Comment by eps1lon** [comment]
I definitely assumed knowledge about the contribution workflow. It seemed patronizing to explain what "request changes" means in GitHub. I'll explain this the next time so that we're on the same level. Sorry about that @n1ru4l. I hope this doesn't stop you from opening issues or commenting on PRs in the future. In my particular github "cultural-bubble" it's generally considered rude to approve/reject PRs targeting code that you do not maintain. So if we accuse each other of toxicity we should consider these different backgrounds.

**Review by ferdaber — APPROVED** [review]
This looks good. We would need to merge in the latest changes in master to get jsx runtime running for the older versions so hopefully no big merge conflicts 🤞

**Review by ferdaber — APPROVED** [review]

**Comment by billyjanitsch** [comment]
I notice that when this merged in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/49701, it kept the `jsx-runtime` declarations in the 16.14 typings but didn't add them in the 17.0 typings. Is that an oversight in the interaction between the two PRs, or is the plan to add them to 17.0 in another PR?

In the meantime, if you use `react@17` with the new JSX runtime, you probably want to install `@types/react@16.14` rather than `@types/react@17`.

**Pull request merged**
